cake/core/src/main/kotlin/org/demoth/cake/modelviewer/Md2ModelLoader.kt

```diff
@@ -66,65 +69,87 @@ class Md2ModelLoader(private val locator: ResourceLocator) {
     ): Md2ShaderModel? {
-        val findModel = locator.findModel(modelName)
-            ?: return null
+        val findModel = locator.findModel(modelName) ?: return null
         val md2Model: Md2Model = readMd2Model(findModel)
 
-        val skins = md2Model.skinNames.map {
+        val embeddedSkins = md2Model.skinNames.map {
             locator.findSkin(it)
         }
-
-/*
-        val modelBuilder = ModelBuilder()
-        modelBuilder.begin()
-*/
-        val modelSkin: ByteArray = if (skins.isNotEmpty()) {
-            skins[skinIndex]
+        val modelSkin: ByteArray = if (embeddedSkins.isNotEmpty()) {
+            embeddedSkins[skinIndex]
         } else {
             if (playerSkin != null) {
                 locator.findSkin(playerSkin)
             } else throw IllegalStateException("No skin found in the model, no player skin provided")
         }
-/*
-        val meshBuilder = modelBuilder.part(
-            "part1",
-            GL_TRIANGLES,
+
+        val diffuse = Texture(PCXTextureData(fromPCX(PCX(modelSkin))))
+
+        val vertexData = buildVertexData(md2Model.glCommands, md2Model.frames)
+
+        val mesh = Mesh(
+            true,
+            vertexData.vertexAttributes.size,
+            vertexData.indices.size,
             VertexAttributes(
-                VertexAttribute.Position(), // 3 floats per vertex, unused by the shader but required by libgdx
-                VertexAttribute.TexCoords(0), // 2 floats per vertex
-                VertexAttribute(VertexAttributes.Usage.Generic, 1, "a_index")
-            ),
-            Material(
-                TextureAttribute(
-                    TextureAttribute.Diffuse,
-                    Texture(PCXTextureData(fromPCX(PCX(modelSkin)))),
-                )
+                VertexAttribute(Generic, 1, "a_vat_index"),
+                TexCoords(1) // in future, normals can also be added here
             )
         )
-*/
-/*
-        val frameVertices = md2Model.getFrameVertices(0)
-        val size = frameVertices.size / 5 // 5 floats per vertex : fixme: not great
-        meshBuilder.addMesh(frameVertices, (0..<size).map { it.toShort() }.toShortArray())
-        val model = modelBuilder.end()
-        val frameBuffers = List(md2Model.frames.size) { i -> md2Model.getFrameVertices(i) }
-*/
-        val meshBuilder = MeshBuilder()
-        meshBuilder.begin(
-            VertexAttributes(
-                VertexAttribute(VertexAttributes.Usage.Generic, 1, "a_index"),
-                VertexAttribute.TexCoords(0) // 2 floats per vertex
+        mesh.setVertices(vertexData.vertexAttributes)
+        mesh.setIndices(vertexData.indices)
+        return Md2ShaderModel(
+            mesh = mesh,
+            vat = createVat(vertexData),
+            diffuse = diffuse,
+        )
+    }
 
-            ),
-            GL20.GL_TRIANGLES
+    private fun createVat(vertexData: Md2VertexData): Texture {
+        return Texture(
+            CustomTextureData(
```

Comment:
todo: add a really deep comment describing the importance of choice of these exact parameters